proof/refine/ARM/StateRelation.thy

```diff
@@ -281,9 +281,29 @@ lemma refills_heads_equal_valid_sched_context':
    \<Longrightarrow> rAmount (refillHd sc') = r_amount (refill_hd sc)
        \<and> rTime (refillHd sc') = r_time (refill_hd sc)"
   apply (frule refills_heads_equal; (solves simp)?)
-  apply (auto simp: sc_relation_def valid_sched_context'_def refillHd_def refills_map_def
-                    refill_map_def wrap_slice_def
-             split: if_splits)
+  by (auto simp: sc_relation_def valid_sched_context'_def refillHd_def refills_map_def
+                 refill_map_def wrap_slice_def
+          split: if_splits)
+
+(* Most sched contexts should satisfy these conditions. The lemma below (standard_sc_common_usage)
+   illustrates a common situation where these conditions hold.
+
+   It seems clear that an abbreviation of this kind is useful, but it's not clear exactly which
+   conditions it should contain. For instance, it does not contain "0 < scRefillMax sc" which one
+   may argue is equally standard.
+*)
+abbreviation (input) standard_sc where
+  "standard_sc sc \<equiv> scRefillMax sc \<le> length (scRefills sc) \<and> scRefillHead sc < scRefillMax sc \<and>
+                    scRefillCount sc \<le> scRefillMax sc \<and> 0 < scRefillCount sc"
```

Comment:
I like this and I have myself encountered situations where I could use something like this. It makes sense to leave it as an abbreviation till we are sure of what conditions it should contain, but we could make it a definition in the future, maybe with a different name, though. “standard_sc” sounds rather too unspecific to me. These are conditions we need for the `wrap_slice` function to make sense, which we use to represent the circular refill list/buffer. Maybe something that features “refills”, like`refills_cond`?

I am starting to think that it would make sense to include these conditions as part of `sc_relation` because the relation between `sc_refills` and `scRefills` won’t hold without them.

Reply:
Well, I guess it’s more of side conditions for valid sc_relation. We can introduce the notion of valid refills for the concrete side.

Reply:
How about `valid_refills'` or `standard_refills`?

Reply:
I vote for `valid_refills’` :)